File: server/frontend-proxy.py

import http.server, socketserver, urllib.request, urllib.parse, os, threading
from http import HTTPStatus

DIST = '/home/ubuntu/celeste-app/dist'

class Handler(http.server.BaseHTTPRequestHandler):
    def log_message(self, *a): pass

    def do_GET(self):
        if self.path.startswith("/api/") or self.path.startswith("/auth/"):
            return self._proxy()
        # Static
        path = self.path.split("?")[0]
        if path == "/" or path == "": path = "/index.html"
        full = os.path.join(DIST, path.lstrip("/"))
        full = os.path.normpath(full)
        if not full.startswith(DIST):
            self.send_error(403); return
        if os.path.isdir(full):
            full = os.path.join(full, "index.html")
        if not os.path.isfile(full):
            full = os.path.join(DIST, "index.html")
        try:
            with open(full, "rb") as f:
                data = f.read()
            ctype = "text/html" if full.endswith(".html") else "application/octet-stream"
            self.send_response(200)
            self.send_header("Content-Type", ctype)
            self.send_header("Content-Length", str(len(data)))
            self.send_header("Cache-Control", "no-store")
            self.end_headers()
            self.wfile.write(data)
        except Exception as e:
            self.send_error(500, str(e))

    def _proxy(self):
        target_url = "http://localhost:3001" + self.path
        try:
            req = urllib.request.Request(target_url, method=self.command, headers={k:v for k,v in self.headers.items()})
            with urllib.request.urlopen(req, timeout=60) as resp:
                body = resp.read()
                self.send_response(resp.status)
                for k, v in resp.headers.items():
                    if k.lower() not in ("transfer-encoding", "connection"):
                        self.send_header(k, v)
                self.send_header("Content-Length", str(len(body)))
                self.end_headers()
                self.wfile.write(body)
        except Exception as e:
            self.send_response(502); self.send_header("Content-Type","text/plain"); self.end_headers()
            self.wfile.write(f"Proxy error: {e}".encode())

    def do_POST(self): self._proxy()
    def do_PUT(self): self._proxy()
    def do_DELETE(self): self._proxy()
    def do_OPTIONS(self): self._proxy()

class Server(socketserver.ThreadingTCPServer):
    allow_reuse_address = True
    daemon_threads = True

# Dual-stack
import socket
class Dual(Server):
    address_family = socket.AF_INET6  # IPv6 + IPv4-mapped IPv6

with Dual(("::", 5174), Handler) as httpd:
    print("Celeste Python proxy on 5174", flush=True)
    httpd.serve_forever()

File: server/frontend-proxy.mjs
// Celeste frontend + API proxy on :5174
// Static files from /home/ubuntu/celeste-app/dist + /api/* proxy to :3001
import http from 'http';
import fs from 'fs';
import path from 'path';

const DIST = '/home/ubuntu/celeste-app/dist';
const BACKEND = 'http://127.0.0.1:3001';
const MIME = {
  '.html':'text/html','.js':'application/javascript','.css':'text/css',
  '.svg':'image/svg+xml','.png':'image/png','.jpg':'image/jpeg','.json':'application/json',
  '.woff2':'font/woff2','.ico':'image/x-icon','.webp':'image/webp',
};

async function proxyApi(req, res) {
  // Forward to backend (raw TCP passthrough)
  const url = req.url;
  const opts = {
    hostname: '127.0.0.1',
    port: 3001,
    path: url,
    method: req.method,
    headers: { ...req.headers, host: 'localhost:3001' },
  };
  const proxyReq = http.request(opts, (proxyRes) => {
    res.writeHead(proxyRes.statusCode, proxyRes.headers);
    proxyRes.pipe(res);
  });
  proxyReq.on('error', (err) => {
    console.error('[proxy err]', err.message);
    res.writeHead(502); res.end('Bad Gateway');
  });
  req.pipe(proxyReq);
}

function serveStatic(req, res) {
  let urlPath = decodeURIComponent(req.url.split('?')[0]);
  if (urlPath === '/' || urlPath === '') urlPath = '/index.html';
  const filePath = path.normalize(path.join(DIST, urlPath));
  if (!filePath.startsWith(DIST)) { res.writeHead(403); return res.end('Forbidden'); }
  fs.readFile(filePath, (err, data) => {
    if (err) {
      // SPA fallback
      fs.readFile(path.join(DIST, 'index.html'), (e2, d2) => {
        if (e2) { res.writeHead(404); return res.end('Not Found'); }
        res.writeHead(200, {'Content-Type':'text/html','Cache-Control':'no-store'});
        res.end(d2);
      });
      return;
    }
    const ext = path.extname(filePath).toLowerCase();
    res.writeHead(200, {
      'Content-Type': MIME[ext] || 'application/octet-stream',
      'Cache-Control': 'no-store',
    });
    res.end(data);
  });
}

const server = http.createServer((req, res) => {
  if (req.url.startsWith('/api/') || req.url.startsWith('/auth/')) {
    return proxyApi(req, res);
  }
  serveStatic(req, res);
});

server.listen(5174, '0.0.0.0', () => {
  console.log('Celeste proxy on 5174 (IPv4 only)');
});
